README.md
boolean3
========

boolean3 represents a substantial re-write of the original boolean package
developed by Bear Braumoeller, Ben Goodrich, and Jacob Kline. This version was
developed by Jason Morgan under the direction of Bear Braumoeller and with
support from The Ohio State University's College of Social and Behavioral
Sciences.


Installation
------------

-   To install the most recent released version of boolean3 from CRAN (this is the
    recommended method):

    ``` r
    install.packages("boolean3")
    ```

-   The latest development version can be installed from github using Hadley
    Wickham's devtools package:

    ``` r
    devtools::install_github("jason-morgan/boolean3")
    ```


NEWS
* Bug Fixes and Changes
** 3.1.7

   + boolboot() now records the convergence status of each
     iteration. Convergence codes follow those found in optimx.

** 3.1.6

   + Fixed a bug in predict() -- function not works as intended.

** 3.1.5

   + Added working examples.

** 3.1.4

   + Change to depend on the parallel package instead of snow.

** 3.1.3

   + Test Mac bootstrap bug.

** 3.1.2

   + Removed MASS from Imports section of DESCRIPTION file.
   + Removed cluster.setup() in do_snow() to eliminate the library() calls.
   + Fix bootstrap sampling. Response was not being sampled, resulting in
     garbage estimates.

** 3.1.1

   + Bumped version requirement for optimx (>= 2013.8.6) to solve some
     compatibility issues.

* New Features
** 3.1.0 (Initial release)

   IMPORTANT NOTE: boolean3 represents a complete re-write of the previous
   boolean package. As a consequence of this, compatibility with previous
   versions of boolean have been broken. The important changes between previous
   releases and boolean3 are listed below.

   + Removed dependence on Zelig.
   + Changed the method of specifying the model to me more consistent with the
     style of other R packages.
   + Greatly improved performance. Many models can be estimated in 1/10th the
     time, or better, when compared to the original boolean package.
   + Added support for the optimizing methods available in the optimx
     package. This includes the ability to apply box constraints, which can be
     useful in maximizing boolean likelihoods.
   + Support for MCMC estimation was dropped.
   + Support for skewed logit (scobit) link functions has been removed (for the
     time being).
   + Added multiprocessor/cluster support for bootstrapping using the parallel
     package.
   + Genetic optimization using rgenoud also provides support for clustering
     with parallel.
   + Plotting of predicted probabilities and likelihood profiles is now done
     with lattice.


tests/testthat/testscript.R
## ----------------------------------------------------------------------------
## Test file for boolean package. VERY ROUGH. UNIT TESTS NEEDED.
## ----------------------------------------------------------------------------

library(devtools)
setwd("~/lib/R")
load_all("boolean3")

## Load the library.
## library(boolean3)
library(optimx)
library(lattice)
library(snow)
library(rlecuyer)
library(mvtnorm)
library(numDeriv)
library(rgenoud)
## library(MCMCpack)
## library(glogis)

## Set up test dataset.
set.seed(12345)
N  <- 2000
Df <- cbind(1, rmvnorm(N, mean=rep(0, 5)))

## Set coefficients
beta.a <- c(-2.00, 0.33, 0.66, 1.00)
beta.b <- c(0.00, 1.50, -0.25)

## Generate path probabilities following a normal model.
y.a <- as.vector(pnorm(tcrossprod(beta.a, Df[, 1:4])))
y.b <- as.vector(pnorm(tcrossprod(beta.b, Df[, c(1, 5, 6)])))

## AND and OR-model
or <- function(x, y) { x + y - x * y }
and <- function(x, y) { x * y }
y.star.OR  <- or(y.a, y.b)
y.star.AND <- and(y.a, y.b)

## Observed responses
y.OR <- rbinom(N, 1, y.star.OR)
y.AND <- rbinom(N, 1, y.star.AND)

## Set up data.frame for estimation
Df <- cbind(1, Df)
Df <- as.data.frame(Df)
Df[,1] <- y.OR
Df[,2] <- y.AND
names(Df) <- c("y.OR", "y.AND", "x1", "x2", "x3", "x4", "x5")


## ----------------------------------------------------------------------------
## Tests of boolean3
## ----------------------------------------------------------------------------

## Set up model, note the new syntax.

## OR model
mod.OR <- boolprep(y.OR ~ (a | b), a ~ x1 + x2 + x3, b ~ x4 + x5,
                   data = Df, family=list(binomial("probit")))

## AND model
mod.AND <- boolprep(y.AND ~ (a & b), a ~ x1 + x2 + x3, b ~ x4 + x5,
                    data = Df, family=list(binomial("probit")))


## The link function used in the model (family = binomial(logit) by default) is
## stored in the boolean object returned by this function. The link function
## now calls built-in C code, which is much faster. A pre-parsed equation for
## the boolean portion of the likelihood is also stored in the boolean
## object. The purpose of storing these in the object is that it facilitates
## easy post-estimation procedures that should always be consistent with the
## original model.
str(mod.OR$link)
str(mod.AND$link)
mod.OR$model$boolmod

## It's also possible to specify a different link function for each path. These
## should be in the same order as specified in the model formula.
mod.OR2 <- boolprep(y.OR ~ (a | b), a ~ x1 + x2 + x3, b ~ x4 + x5,
                    data = Df,
                    family=list(binomial("probit"), binomial("logit")))
str(mod.OR2$link)

## Fit the model using the nlminb optimizer.
(fit.OR <- boolean(mod.OR, method="nlminb", control=list(trace=1)))


cbind(est=coef(fit.OR), truth=c(beta.a, beta.b))

## Fit with the nlminb and nlm optimizer.
(fit1.OR <- boolean(mod.OR, method="nlm"))

## Re-fit, with BFGS and a higher maximum iterations. All of the options that
## go along with nlm(), optim(), and genoud() should be transparently passable.
(fit2.OR <- boolean(mod.OR, method="BFGS", control = list(maxit = 500)))

## Induce a convergence warning message.
(fit3.OR <- boolean(mod.OR, method="BFGS", control = list(maxit = 5)))

## Using MCMC. Then take a more detailed look at the results using tools from
## coda package. BROKEN SPECIFICATION
## (fit4.OR <- boolean(mod.OR, method="mcmc", mcmc=30000))

## Using all methods provided by optimx, increase verbosity. MCMC, SANN, and
## genoud are not used when all.methods=TRUE.
## (fit5.OR <- boolean(mod.OR, control=list(trace=0, all.methods=TRUE)))

## Using genoud.
(fit6.OR <- boolean(mod.OR, method="genoud", cluster=c("localhost", "localhost"),
                    print.level=2))
(fit6.OR <- boolean(mod.OR, method="genoud", print.level=2))

## Using SANN.
(fit7.OR <- boolean(mod.OR, method="SANN"))

## The fit is stored as "model.fit", within the boolean object. The Hessian is
## always saved and must be inverted to get the covariance matrix (the summary
## function below does this for you). *Important note*: all optimizers are set
## to *minimize* the negative of the log-likelihood. This is the preferred
## method for many optimizers, so it was easiest simply to calc -1*logLik at
## presentation. This does, however, affect the sign of the Hessian saved in the
## model.fit object; i.e., you don't want to multiply by -1 when calculating
## standard errors. If this becomes an issue, I can simply store the Hessian
## returned by the optimizers as -1 * Hessian.
str(fit.OR$model.fit)

## Create a summary object, saving and printing it. Then take a look at the
## objects stored in the summary object.
(smry <- summary(fit.OR))
str(smry)

## Extract log-likelihood and coefficient vector.
logLik(fit.OR)
coef(fit.OR)
smry$methods$nlminb$coef

## Now extract profile likelihoods (actually, estimated likelihoods, if I
## understand the likelihood theory -- need to check this) for all of the
## covariates. Since the plot is a lattice object, it can be stored as well. The
## default is to simply plot it, but storing it allows alternative plots to be
## made without recalculating the profiles.
(prof <- boolprof(fit.OR))

## Now, extract the plots for x1_a and x4_b.
plot(prof, y = c("x1_a", "x4_b"))
plot(prof, y = c(1, 3), scales = list(y = list(relation = "free")))

## You can use variable or index matching with boolprof to select particular
## covariates of interest.
boolprof(fit.OR, vars = c(1, 3))
boolprof(fit.OR, vars = c("x1_a", "x4_b"))

## Plots of predicted probabilities are available with boolprob. With boolprob,
## either vars or newdata *must* be specified.
boolprob(fit.OR, vars = c("x1_a", "x4_b"))
boolprob(fit.OR, vars = c(2, 3, 4, 6))
boolprob(fit.OR, vars = c(2:4, 6, 7))

## Specifying conf.int = TRUE gets you simulated confidence intervals. The
## number of samples to pull from the distribution of the estimated
## coefficients is controlled by n, which n=100 is default.
(prob <- boolprob(fit.OR, vars = c(2, 3, 4, 6), n = 100, conf.int = TRUE))

## Choose a different method estimate upon which to base the estimates.
(prob <- boolprob(fit.OR, method="spg", vars=c(2, 3, 4, 6), n=100,
                  conf.int=TRUE))

## As with the other components of the model, you can extract the predicted
## probabilities.
str(prob)
prob$est

## Inspect MCMC mixing.
plot(fit4$model.fit$mcmc$mcmc)

## Bootstrapping models.
(bs <- boolboot(fit.OR, n=10))

b <- coef(fit.OR)
se <- summary(fit.OR)$methods$nlminb$coef[,2]
lower <- b - 4*se
upper <- b + 4*se
cbind(lower, upper)

(bs <- boolboot(fit.OR, n=10))
cbind(lower, upper)

summary(fit.OR)

(bs <- boolboot(fit.OR, n=20, cluster=c("localhost", "localhost")))


R/bootstrap.R
### ----------------------------------------------------------------------------
### This file is part of boolean3
###
### Copyright (C) 2011--2014 Jason W. Morgan <morgan.746@osu.edu>
###
### boolean3 represents a substantial re-write of the original boolean package
### developed by Bear Braumoeller, Ben Goodrich, and Jacob Kline. This version
### was developed under the direction of Bear Braumoeller and with support from
### The Ohio State University's College of Social and Behavioral Sciences.
###
### boolean3 is free software: you can redistribute it and/or modify it under
### the terms of the GNU General Public License as published by the Free
### Software Foundation, either version 3 of the License, or (at your option)
### any later version.
###
### This program is distributed in the hope that it will be useful, but WITHOUT
### ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
### FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
### more details.
###
### You should have received a copy of the GNU General Public License along with
### this program.  If not, see <http://www.gnu.org/licenses/>.
###
### ----------------------------------------------------------------------------

##' Performs bootstrap estimates of a boolean model.
##'
##' \code{boolboot} performs bootstrap estimated of a boolean model specified
##' by \code{\link{boolprep}} and first estimated with \code{\link{boolean}}.
##' @title Bootstrap for Boolean Models
##' @param obj boolean model object as produced by \code{\link{boolprep}} and
##' first estimated with \code{\link{boolean}}.
##' @param n integer specifying the number of bootstrap estimates. Defaults to
##' \code{100}.
##' @param method string specifying the method of estimation. The specified
##' method should be one of those available from the \code{\link{optimx}} or
##' \code{\link{optim}} functions. Defaults to \code{"nlminb"}.
##' @param cluster string vector specifying hosts to use for parallel processing
##' through \code{parallel} (see \code{\link{makeCluster}}). Defaults to \code{NULL}
##' indicating no clustering.
##' @param ... additional parameters to pass on to subsequent functions.
##' @return \code{boolboot} returns a \code{boolboot} model object. This
##' object is identical to a boolean model object but with an additional
##' \code{model.boot} slot containing the results of the bootstrap. A separate
##' object type is used to help prevent the accidential loss of bootstrap
##' estimates.
##' @author Jason W. Morgan (\email{morgan.746@@osu.edu})
##' @references Braumoeller, Bear F. (2003) ``Causal Complexity and the Study
##' of Politics.'' \emph{Political Analysis} 11(3): 209--233.
##' @seealso See \code{\link{boolprep}} for model setup, \code{\link{boolean}}
##' for estimation, the \code{parallel} package for details on clustering,
##' and \code{\link{optimx}} and \code{\link{optim}} for estimation methods.
##' @export
boolboot <- function(obj, n=100, method="nlminb", cluster=NULL, ...)
{
    if (!inherits(obj, "boolean"))
        stop("boolboot requires a boolean object as outputed by boolprep", call.=FALSE)
    if (inherits(obj, "boolboot"))
        stop("model object already contains bootstrapping results. Halting to prevent data loss", call.=FALSE)

    ## A single method has to be specified. Defaults to nlminb
    if (length(method) > 1)
        obj$method <- method[1]
    else
        obj$method <- method

    if (!is.null(cluster)) {
        ## Bootstrap the model with parallel cluster.
        bs_closure <- bs_mod(obj, ...)
        Results <- do_parallel(cluster, bs_closure, n, ...)
        print(Results)
        ## Results <- t(do.call(rbind, Results))
        par <- sapply(Results, `[[`, "par")
        conv <- sapply(Results, `[[`, "conv")
        Results <- list(par=par, conv=conv)
    }
    else {
        ## Bootstrapping without cluster.
        bs_closure <- bs_mod(obj, ...)
        Results <- lapply(1:n, bs_closure, ...)
        par <- sapply(Results, `[[`, "par")
        conv <- sapply(Results, `[[`, "conv")
        Results <- list(par=par, conv=conv)
    }

    obj[["model.boot"]] <- list(coef=t(Results$par), conv=Results$conv,
                                samples=n)
    class(obj) <- c("boolboot", class(obj))
    obj
}

do_parallel <- function(cluster, fn, n, ...)
{
    Cluster <- makeCluster(cluster, type="SOCK")
    clusterSetRNGStream(Cluster)
    Results <- clusterApplyLB(Cluster, 1:n, fn, ...)
    stopCluster(Cluster)
    Results
}

bs_sample <- function(obj, ...)
{
    ## Samples with replacement from the model.matrix of a boolean model object.
    S <- model.matrix(obj)
    Y <- obj$response
    idx <- sample(1:nrow(S), replace=TRUE)
    list(y=Y[idx], samp=S[idx,])
}

bs_obj <- function(obj, ...)
{
    ## Returns an augmented model object that includes the new bootstrapped
    ## sample.
    tmp <- bs_sample(obj)
    new.y <- tmp$y
    new.samp <- tmp$samp

    for (mod in names(obj$coef.idx)) {
        nam <- colnames(obj$frame[[mod]])
        obj$frame[[mod]] <- new.samp[, obj$coef.idx[[mod]]]
        colnames(obj$frame[[mod]]) <- nam
    }
    obj$response <- new.y
    obj$calc_lik <- make_lik(obj)
    obj
}

bs_mod <- function(obj, ...)
{
    ## Function used internally for bootstrapping a model. Returns a closure.
    n <- 1
    function(...) {
        cat("Bootstrap interation:", n, "\n")
        n <<- n+1
        bs_fit <- boolean(bs_obj(obj), ...)
        list(par=bs_fit$model.fit[[1]]$par,
             conv=bs_fit$model.fit[[1]]$conv)
    }
}


R/boolean-package.R
## ----------------------------------------------------------------------------
## This file is part of boolean3
##
## Copyright (C) 2011--2014 Jason W. Morgan <morgan.746@osu.edu>
##
## boolean3 represents a substantial re-write of the original boolean package
## developed by Bear Braumoeller, Ben Goodrich, and Jacob Kline. This version
## was developed under the direction of Bear Braumoeller and with support from
## The Ohio State University's College of Social and Behavioral Sciences.
##
## boolean3 and is free software: you can redistribute it and/or modify it under
## the terms of the GNU General Public License as published by the Free Software
## Foundation, either version 3 of the License, or (at your option) any later
## version.
##
## This program is distributed in the hope that it will be useful, but WITHOUT
## ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
## FOR A PARTICULAR PURPOSE.  See the GNU General Public License for more
## details.
##
## You should have received a copy of the GNU General Public License along with
## this program.  If not, see <http://www.gnu.org/licenses/>.
##
## ----------------------------------------------------------------------------

##' @import stats rgenoud numDeriv optimx parallel mvtnorm rlecuyer lattice
NULL

##'
##' Boolean binary response models are a family of partial-observability binary
##' response models designed to permit researchers to model causal complexity,
##' or multiple causal ``paths'' to a given outcome.
##'
##' Boolean permits estimation of Boolean binary response models (see
##' Braumoeller 2003 for derivation), which are a family of
##' partial-observability n-variate models designed to permit researchers to
##' model causal complexity, or multiple causal ``paths'' to a given
##' outcome. The various ``paths'' are modeled as latent dependent variables
##' that are multiplied together in a manner determined by the logic of their
##' (Boolean) interaction. If, for example, we wanted to model a situation in
##' which diet OR smoking causes heart failure, we would use one set of
##' independent variables (caloric intake, fat intake, etc.) to predict the
##' latent probability of diet-related coronary failure (\eqn{y1^*}), use
##' another set of variables (cigarettes smoked per day, exposure to second-hand
##' smoke, etc.)  to predict the latent probability of smoking-related coronary
##' failure (\eqn{y2^*}), and model the observed outcome (\eqn{y}, or coronary
##' failure) as a function of the Boolean interaction of the two: \eqn{\Pr(y=1) =
##' 1-([1-y1^*] \times [1-y2^*])}. Independent variables that have an impact on both
##' latent dependent variables can be included in both paths. Any combination of
##' any number of ANDs and ORs can be modeled, though the procedure becomes
##' exponentially more data-intensive as the number of interactions increases.
##'
##' \tabular{ll}{
##' Package: \tab boolean3 \cr
##' Type: \tab Package\cr
##' Version: \tab 3.1.7\cr
##' Date: \tab 2015-09-24\cr
##' License: \tab GPL-3 \cr
##' LazyLoad: \tab yes\cr
##' }
##'
##' @name boolean3-package
##' @aliases boolean3
##' @docType package
##' @title Modeling Causal Complexity
##' @note \code{boolean3} was developed by Jason W. Morgan under the direction
##' of Bear Braumoeller with support from The Ohio State University's College of
##' Social and Behavioral Sciences. The package represents a significant
##' re-write of the original \code{boolean} implementation developed by Bear
##' Braumoeller, Ben Goodrich, and Jacob Kline. Please see the release notes and
##' accompanying documentation for details regarding the many changes made in
##' this version.
##' @author Jason W. Morgan (\email{morgan.746@@osu.edu})
##' @references Braumoeller, Bear F. (2003) ``Causal Complexity and the Study of
##' Politics.'' \emph{Political Analysis} 11(3): 209--233.
##' @keywords package
##' @seealso See \code{\link{boolprep}} for model setup and
##' \code{\link{boolean}} for estimation.
##' @examples
##' ## Generate some fake data
##' require(mvtnorm)
##' set.seed(12345)
##' N  <- 2000
##' Df <- cbind(1, rmvnorm(N, mean=rep(0, 5)))
##'
##' ## Set coefficients
##' beta.a <- c(-2.00, 0.33, 0.66, 1.00)
##' beta.b <- c(0.00, 1.50, -0.25)
##'
##' ## Generate path probabilities following a normal model.
##' y.a <- as.vector(pnorm(tcrossprod(beta.a, Df[, 1:4])))
##' y.b <- as.vector(pnorm(tcrossprod(beta.b, Df[, c(1, 5, 6)])))
##'
##' ## AND and OR-model
##' or <- function(x, y) { x + y - x * y }
##' and <- function(x, y) { x * y }
##' y.star.OR  <- or(y.a, y.b)
##' y.star.AND <- and(y.a, y.b)
##'
##' ## Observed responses
##' y.OR <- rbinom(N, 1, y.star.OR)
##' y.AND <- rbinom(N, 1, y.star.AND)
##'
##' ## Set up data.frame for estimation
##' Df <- cbind(1, Df)
##' Df <- as.data.frame(Df)
##' Df[,1] <- y.OR
##' Df[,2] <- y.AND
##' names(Df) <- c("y.OR", "y.AND", "x1", "x2", "x3", "x4", "x5")
##'
##' ## Before estimating, boolean models need to be specified using the
##' ## boolprep function.
##'
##' ## OR model, specified to use a probit link for each causal path. This
##' ## matches the data generating process above.
##' mod.OR <- boolprep(y.OR ~ (a | b), a ~ x1 + x2 + x3, b ~ x4 + x5,
##'                    data = Df, family=list(binomial("probit")))
##'
##' ## IF you really want to, it's also possible to specify a different
##' ## link function for each causal path. These should be in the same
##' ## order as specified in the model formula.
##' mod.OR2 <- boolprep(y.OR ~ (a | b), a ~ x1 + x2 + x3, b ~ x4 + x5,
##'                     data = Df, family=list(binomial("probit"),
##'                                    binomial("logit")))
##'
##' ## Fit the prepared model using the nlminb optimizer (the default).
##' (fit.OR <- boolean(mod.OR, method="nlminb", control=list(trace=1)))
##'
##' ## Multiple optimizers can be specified in a single call to boolean.
##' ## Here we fit with the nlm and nlminb optimizers.
##' (fit1.OR <- boolean(mod.OR, method=c("nlm", "nlminb")))
##'
##' ## Re-fit, with BFGS and a higher maximum number of iterations. All
##' ## of the options that go along with nlm(), optim(), and genoud() should
##' ## be transparently passable through boolean.
##' (fit2.OR <- boolean(mod.OR, method="BFGS", control = list(maxit = 500)))
##'
##' ## Induce a convergence warning message.
##' (fit3.OR <- boolean(mod.OR, method="BFGS", control = list(maxit = 5)))
##'
##' \dontrun{
##' ## Now estimate model with genoud, a genetic optimizer. This has the
##' ## capability of using multiple processors via parallel.
##' (fit6.OR <- boolean(mod.OR, method="genoud",
##'                     cluster=c("localhost", "localhost"),
##'                     print.level=2))
##'
##' ## The default SANN optimizer is also available.
##' (fit7.OR <- boolean(mod.OR, method="SANN"))
##' }
##'
##' ## The fit is stored as "model.fit", within the boolean object.
##' str(fit.OR$model.fit)
##'
##' ## Create a summary object, saving and printing it. Then take a look at
##' ## the objects stored in the summary object.
##' (smry <- summary(fit.OR))
##' str(smry)
##'
##' ## Extract log-likelihood and coefficient vector.
##' logLik(fit.OR)
##' coef(fit.OR)
##'
##' \dontrun{
##' ## Display the contours of the likelihood given a change the value of
##' ## the coefficients. Despite the function name, these are not true
##' ## profile likelihoods as they hold all other coefficients fixed at
##' ## their MLE.
##' (prof <- boolprof(fit.OR))
##'
##' ## Extract the plots for x1_a and x4_b.
##' plot(prof, y = c("x1_a", "x4_b"))
##' plot(prof, y = c(1, 3), scales = list(y = list(relation = "free")))
##'
##' ## You can also use variable or index matching with boolprof to select
##' ## particular covariates of interest.
##' boolprof(fit.OR, vars = c(1, 3))
##' boolprof(fit.OR, vars = c("x1_a", "x4_b"))
##'
##' ## Plots of predicted probabilities are available through boolprob.
##' ## With boolprob, either vars or newdata *must* be specified.
##' boolprob(fit.OR, vars = c("x1_a", "x4_b"))
##' boolprob(fit.OR, vars = c(2, 3, 4, 6))
##'
##' ## Specifying conf.int = TRUE produces simulated confidence intervals.
##' ## The number of samples to pull from the distribution of the estimated
##' ## coefficients is controlled by n; n=100 is default. This can take a
##' ## while.
##' (prob <- boolprob(fit.OR, vars = c(2, 3, 4, 6), n = 1000,
##'                   conf.int = TRUE))
##'
##' ## Choose a different method estimate upon which to base the estimates.
##' (prob <- boolprob(fit1.OR, method="nlm", vars=c(2, 3, 4, 6), n=1000,
##'                   conf.int=TRUE))
##'
##' ## As with the other components of the model, you can extract the
##' ## predicted probabilities.
##' str(prob)
##' prob$est
##'
##' ## Bootstrapping is also possible, and is the recommended method of
##' ## making inferences. The boolbool function uses a simple sampling scheme:
##' ## it resamples repeatedly from the provided data.frame. More the complex
##' ## data structures with, for example, clustering, need to be dealt with
##' ## manually.
##' (bs <- boolboot(fit, n=10))
##'
##' ## boolboot supports bootstrapping across multiple processors.
##' (bs <- boolboot(fit, n=100, cluster=c("localhost", "localhost")))
##'
##' }
NULL
